docs/design/2023-04-11-dist-task.md

```diff
@@ -0,0 +1,185 @@
+| step              | INT(11)                                        |      |
+
+`Global subtask queue`
+
+| Name              | Type                                       | Memo |
+|-------------------|--------------------------------------------|------|
+| id                | bigint not null auto_increment primary key |      |
+| namespace         | varchar(256)                               |      |
+| task_key          | varchar(256)                               |      | 
+| ddl_physical_tid  | bigint(20)                                 |      |
+| type              | int                                        |      |  
+| exec_id           | varchar(256)                               |      |
+| exec_expired      | timestamp                                  |      |
+| state             | varchar(64) not null                       |      |
+| checkpoint        | longblob not null                          |      | 
+| start_time        | bigint                                     |      |        
+| state_update_time | bigint                                     |      | 
+| meta              | longblob                                   |      |                                            |      |
+
+### Interface
+
+Dispatcher
+```go
+// Dispatch defines the interface for operations inside a dispatcher.
+type Dispatch interface {
+	// Start enables dispatching and monitoring mechanisms.
+	Start()
+	// GetAllSchedulerIDs gets handles the task's all available instances.
+	GetAllSchedulerIDs(ctx context.Context, gTaskID int64) ([]string, error)
+	// Stop stops the dispatcher.
+	Stop()
+}
+
+type dispatcher struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	taskMgr *storage.TaskManager
+	wg      tidbutil.WaitGroupWrapper
+	gPool   *spool.Pool
+
+	runningGTasks struct {
+		syncutil.RWMutex
+		taskIDs map[int64]struct{}
+	}
+	detectPendingGTaskCh chan *proto.Task
+}
+```
+Scheduler
+```go
+// InternalScheduler defines the interface of an internal scheduler.
+type InternalScheduler interface {
+	Start()
+	Stop()
+	Run(context.Context, *proto.Task) error
+	Rollback(context.Context, *proto.Task) error
+}
+
+// Scheduler defines the interface of a scheduler.
+// User should implement this interface to define their own scheduler.
+type Scheduler interface {
+	InitSubtaskExecEnv(context.Context) error
+	SplitSubtask(subtask []byte) []proto.MinimalTask
+	CleanupSubtaskExecEnv(context.Context) error
+	Rollback(context.Context) error
+}
+
+// InternalSchedulerImpl is the implementation of InternalScheduler.
+type InternalSchedulerImpl struct {
+	ctx       context.Context
+	cancel    context.CancelFun
+	id        string
+	taskID    int64
+	taskTable TaskTable
+	pool      Pool
+	wg        sync.WaitGroup
+	logCtx    context.Context
+
+	mu struct {
+		sync.RWMutex
+		err error
+		// runtimeCancel is used to cancel the Run/Rollback when error occurs.
+		runtimeCancel context.CancelFunc
+	}
+}
```

Comment:
Maybe we don't need to mention the implementation details here.

Reply:
Only interface definition.